python/ray/autoscaler/updater.py

```diff
@@ -120,31 +122,51 @@ def sync_file_mounts(self, sync_cmd):
                 "~/ray_bootstrap_key.pem", "~/ray_bootstrap_config.yaml"
             ]
 
+        def do_sync(remote_path, local_path, allow_non_existing_paths=False):
+            if allow_non_existing_paths and not os.path.exists(local_path):
+                # Ignore missing source files. In the future we should support
+                # the --delete-missing-args command to delete files that have
+                # been removed
+                return
+
+            assert os.path.exists(local_path), local_path
+
+            if os.path.isdir(local_path):
+                if not local_path.endswith("/"):
+                    local_path += "/"
+                if not remote_path.endswith("/"):
+                    remote_path += "/"
+
+            with LogTimer(self.log_prefix +
+                          "Synced {} to {}".format(local_path, remote_path)):
+                self.cmd_runner.run("mkdir -p {}".format(
+                    os.path.dirname(remote_path)))
+                sync_cmd(local_path, remote_path)
+
+                if remote_path not in nolog_paths:
+                    # todo: timed here?
+                    cli_logger.print("{} from {}", cf.bold(remote_path),
+                                     cf.bold(local_path))
+
         # Rsync file mounts
         with cli_logger.group(
-                "Processing file mounts", _numbered=("[]", 2, 5)):
+                "Processing file mounts", _numbered=("[]", 2, 6)):
             for remote_path, local_path in self.file_mounts.items():
-                assert os.path.exists(local_path), local_path
-                if os.path.isdir(local_path):
-                    if not local_path.endswith("/"):
-                        local_path += "/"
-                    if not remote_path.endswith("/"):
-                        remote_path += "/"
-
-                with LogTimer(self.log_prefix + "Synced {} to {}".format(
-                        local_path, remote_path)):
-                    self.cmd_runner.run("mkdir -p {}".format(
-                        os.path.dirname(remote_path)))
-                    sync_cmd(local_path, remote_path)
-
-                    if remote_path not in nolog_paths:
-                        # todo: timed here?
-                        cli_logger.print("{} from {}", cf.bold(remote_path),
-                                         cf.bold(local_path))
+                do_sync(remote_path, local_path)
+
+        if self.worker_file_mounts:
+            with cli_logger.group(
+                    "Processing worker file mounts", _numbered=("[]", 3, 6)):
+                for remote_path, local_path in self.worker_file_mounts.items():
+                    do_sync(
+                        remote_path, local_path, allow_non_existing_paths=True)
```

Comment:
This doesn't work if you're starting a cluster right? Your local laptop wouldn't have "worker file mounts"?

Reply:
the create_cluster flow does not set worker_file_mounts so it wont try to sync worker_file_mounts from the local laptop to the head node.

When syncing from the head node to the worker nodes, the `allow_non_existing_paths` param makes it so it wont error out if the files are not yet in the head node yet. It will just start syncing once those files exist.

Reply:
OK got it - is this expected to be user-facing/should the user be setting this?

Reply:
no that param is only for the private function. Its completely internal.

Reply:
Sorry I meant, are `worker_file_mounts` user-facing/should the user be setting it?

Reply:
ah yes, the user should be setting it if they want to do any sort of syncing of head_node to worker nodes. One example is if they are working in the head node and modify some project files and want those project files to be updated in the worker nodes automatically

Reply:
Got it. I think it'd be better to make this somehow explicit that this is only head<->worker, and not laptop->head. Right now, "worker_file_mounts" seems to suggest that the behavior will be similar to standard file_mounts.

Something like "cluster_node_sync_files", indicating that the listed files will be kept in-sync across the cluster, but excludes any files on the laptop (perhaps some behavior will need to be changed in order to enable this).

Reply:
okay, i'll name it `cluster_synced_files`